arch/risc-v/src/common/riscv_pthread_start.c

```diff
@@ -66,11 +66,17 @@
 void up_pthread_start(pthread_trampoline_t startup,
                       pthread_startroutine_t entrypt, pthread_addr_t arg)
 {
+#ifdef CONFIG_ARCH_USE_S_MODE
+  /* Let ksys_call3() do all of the work */
+
+  ksys_call3(SYS_pthread_start, (uintptr_t)startup, (uintptr_t)entrypt,
+             (uintptr_t)arg);
+#else
   /* Let sys_call3() do all of the work */
 
   sys_call3(SYS_pthread_start, (uintptr_t)startup, (uintptr_t)entrypt,
```

Comment:
@xiaoxiang781216 would it be more logical if I renamed e.g. the call to sys_call3 to ksys_call3 here, and then did the implementation of that differently by flagging with CONFIG_ARCH_USE_S_MODE.

Because to me it still makes sense to keep two separate functions, one is called by kernel code, other is called by user code via e.g. libc. So in the future, sys_callx would be used by user side only, and ksys_callx would be for kernel ?

Reply:
There are two approach here:

1. Define two functions sys_callx and ksys_callx, but we need guard the prototype and implementation by the related CONFIG_XXX to ensure the user and kernel library don't reference the wrong symbol in the compile and link.
2. Unify to the same function sys_callx which exist in both user/kernel space, but change the implementation accordingly.

The second approach is better since all caller can automatically work in both user space and kernel space without any change. Please reconsider how malloc fix the similar issue which is very elegant and extensible.

Reply:
Citation from kmalloc.h
```

/* For a monolithic, kernel-mode NuttX build.  Special allocators must be
 * used.  Otherwise, the standard allocators prototyped in stdlib.h may
 * be used for both the kernel- and user-mode objects.
 */

/* This family of allocators is used to manage user-accessible memory
 * from the kernel.  In the flat build, the following are declared in
 * stdlib.h and are directly callable.  In the kernel-phase of the kernel
 * build, the following are defined in userspace.h as macros that call
 * into user-space via a header at the beginning of the user-space blob.
 */

#define kumm_initialize(h,s)     umm_initialize(h,s)
#define kumm_addregion(h,s)      umm_addregion(h,s)

#define kumm_calloc(n,s)         calloc(n,s);
#define kumm_malloc(s)           malloc(s)
#define kumm_malloc_size(p)      malloc_size(p)
```

So there are indeed separate prototypes for the kernel to access user space. And below the mapping of kernel heap is done to the flat mode combined/shared heap, or to a dedicated kernel heap which is required in CONFIG_BUILD_KERNEL.

```
/* This family of allocators is used to manage kernel protected memory */

#ifndef CONFIG_MM_KERNEL_HEAP
/* If this the kernel phase of a kernel build, and there are only user-space
 * allocators, then the following are defined in userspace.h as macros that
 * call into user-space via a header at the beginning of the user-space blob.
 */

#  define kmm_initialize(h,s)    /* Initialization done by kumm_initialize */
#  define kmm_addregion(h,s)     umm_addregion(h,s)

#  define kmm_calloc(n,s)        calloc(n,s);
#  define kmm_malloc(s)          malloc(s)
#  define kmm_malloc_size(p)     malloc_size(p)
#  define kmm_zalloc(s)          zalloc(s)
#  define kmm_realloc(p,s)       realloc(p,s)
#  define kmm_memalign(a,s)      memalign(a,s)
#  define kmm_free(p)            free(p)
#  define kmm_mallinfo()         mallinfo()

#else
/* Otherwise, the kernel-space allocators are declared in
 * include/nuttx/mm/mm.h and we can call them directly.
 */

#endif
```

I'm probably missing something, but if sys_callx is the only declaration, how do we ensure this does not happen ?

> Otherwise, the standard allocators prototyped in stdlib.h may be used for both the kernel- and user-mode objects.